AssemblerLib/VectorMatrixAssembler.h

```diff
@@ -49,15 +57,41 @@ class VectorMatrixAssembler
     {
         assert(_data_pos.size() > id);
 
-        LocalToGlobalIndexMap::RowColumnIndices const& indices = _data_pos[id];
+        std::vector<GlobalIndexType> indices;
+
+        // Local matrices and vectors will always be ordered by component,
+        // no matter what the order of the global matrix is.
+        for (unsigned c=0; c<_data_pos.getNumComponents(); ++c)
+        {
+            auto const& idcs = _data_pos(id, c).rows;
+            indices.reserve(indices.size() + idcs.size());
+            indices.insert(indices.end(), idcs.begin(), idcs.end());
+        }
+
+        std::vector<double> localX;
+        std::vector<double> localX_pts;
+
+        if (_x)         localX.reserve(indices.size());
+        if (_x_prev_ts) localX_pts.reserve(indices.size());
+
+        for (auto i : indices)
+        {
+            if (_x)         localX.emplace_back(_x->get(i));
```

Comment:
Get values entry by entry is slow in PETSc. There is a member in class PETScVector called  
```
get(const std::vector<PetscInt> &e_idxs,  T_SUBVEC &sub_vec)
```
You can try to perform it like
```
_x->get(indices, localX);
```
For this purpose, you need to add        
```
 template<class T_SUBVEC> void get(const std::vector<[Int Type of the owner package]> &e_idxs,
                                          T_SUBVEC &sub_vec)
```
to all vector classes in MathLib/LinAlg. 

Besides, do not forget to set  indices be positive.